### ISSUE TITLE
[vcpkg tool] vcpkg behind proxy

### ISSUE THREAD

**Issue by AndreyAlifanov** [microsoft/vcpkg#41987]
Hello.

We have corporate proxy with exceptions for our intranet. vcpkg works improperly in this configuration.
git operations work as excpected. but vcpkg itself can't download archives from our internal git repositories. it tries to load them via proxy, but it should use exceptions, and not try go out.

**Environment**
- OS: Windows

**To Reproduce**
```
...
1>-- Extracting source E:/Tools/vcpkg/downloads/openssl-d82e959e621a3d597f1e0d50ff8c2d8b96915fd7.tar.gz
1>-- Using source at E:/Tools/vcpkg/buildtrees/openssl/src/8b96915fd7-66e150de77.clean
1>-- Downloading https://vcpkg.ourcompany.com/StrawberryPerl/Perl-Dist-Strawberry/releases/download/SP_5380_5361/strawberry-perl-5.38.0.1-64bit-portable.zip -> strawberry-perl-5.38.0.1-64bit-portable.zip...
...
1>EXEC : error : Missing strawberry-perl-5.38.0.1-64bit-portable.zip and downloads are blocked by x-block-origin.
1>EXEC : error : https://vcpkg.ourcompany.com/StrawberryPerl/Perl-Dist-Strawberry/releases/download/SP_5380_5361/strawberry-perl-5.38.0.1-64bit-portable.zip: failed: status code 503
1>[DEBUG] E:\Tools\vcpkg-tool\src\vcpkg\base\downloads.cpp(1041):
...
1>
1>CMake Error at scripts/cmake/vcpkg_download_distfile.cmake:32 (message):
1>
1>EXEC : Failed to download file with error : 1
1>      If you are using a proxy, please check your proxy setting. Possible causes are:
1>
1>      1. You are actually using an HTTP proxy, but setting HTTPS_PROXY variable
1>         to `https://address:port`. This is not correct, because `https://` prefix
1>         claims the proxy is an HTTPS proxy, while your proxy (v2ray, shadowsocksr
1>         , etc..) is an HTTP proxy. Try setting `http://address:port` to both
1>         HTTP_PROXY and HTTPS_PROXY instead.
1>
1>      2. If you are using Windows, vcpkg will automatically use your Windows IE Proxy Settings
1>         set by your proxy software. See https://github.com/microsoft/vcpkg-tool/pull/77
1>         The value set by your proxy might be wrong, or have same `https://` prefix issue.
1>
1>      3. Your proxy's remote server is out of service.
1>
1>      If you've tried directly download the link, and believe this is not a temporary
1>      download server failure, please submit an issue at https://github.com/Microsoft/vcpkg/issues
1>      to report this upstream download server failure.
1>
1>
1>Call Stack (most recent call first):
1>  scripts/cmake/vcpkg_download_distfile.cmake:270 (z_vcpkg_download_distfile_show_proxy_and_fail)
1>  scripts/cmake/vcpkg_find_acquire_program.cmake:170 (vcpkg_download_distfile)
1>  C:/Users/.../AppData/Local/vcpkg/registries/git-trees/9c4da591bd5906fe8631a2815353db45f69e1a9b/portfile.cmake:13 (vcpkg_find_acquire_program)
1>  scripts/ports.cmake:172 (include)
```
**Expected behavior**
vcpkg-tool should use proxy properly and respect exceptions for intranet.

**Comment by AndreyAlifanov** [comment]
I found possible error in vcpkg-tool.
In file src/vcpkg/base/downloads.cpp the next code snippet exists:
```
auto h = WinHttpOpen(
           L"vcpkg/1.0", WINHTTP_ACCESS_TYPE_NO_PROXY, WINHTTP_NO_PROXY_NAME, WINHTTP_NO_PROXY_BYPASS, 0);
```
after this code an attempt to setup proxy exists. but it works very strange, it ignores bypass settings.
and seems as unnecessary code.

I've changed it with next code:
```
auto h = WinHttpOpen(
           L"vcpkg/1.0", WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY, WINHTTP_NO_PROXY_NAME, WINHTTP_NO_PROXY_BYPASS, 0);
```
and removed manual setup of proxy below.

now my functionality works properly, as I see. but I'm not sure if my changes are correct in all cases.

**Comment by MonicaLiu0311** [comment]
@BillyONeal Could you please help take a look at this issue?

**Comment by AndreyAlifanov** [comment]
Why issue closed?
It's a real problem in using vcpkg behind proxy with local addresses bypass.
And it's because of ancient code in vcpkg, related to Windows XP/7/8.

**Comment by AndreyAlifanov** [comment]
> At the moment we intentionally disable WinHTTP's WPAD support because several of our download ops use curl.exe, and curl.exe does not implement WPAD. You can workaround this by configuring your proxy using the HTTP_PROXY or HTTPS_PROXY environment variables.
> 

We can't use proxy bypass list with these settings. Here is the initial problem. Respecting NO_PROXY environment variable would be good solution. But now vcpkg for Windows ignores it.

And, by the way, code from my post works like a charm in fork. So, why don't you try use it?

**Comment by BillyONeal** [comment]
> We can't use proxy bypass list with these settings.

I'm not sure I understand. What is there to bypass given that we have it set to no proxy in the first place?

> Here is the initial problem. Respecting NO_PROXY environment variable would be good solution. But now vcpkg for Windows ignores it.

I'm sure we would be happy to accept a PR that makes this be respected as long as it works the same for both the curl and WinHTTP backends.

> And, by the way, code from my post works like a charm in fork. So, why don't you try use it?

Because `WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY` will engage WPAD and thus trigger a meaningful / observable difference based on whether the curl backend or the WinHTTP backend are selected. Given that we don't really give users control over which backend is selected, that could be bad. (We basically try to use WinHTTP unless we are doing something the WinHTTP backend doesn't know how to do)

**Comment by AndreyAlifanov** [comment]
Thank you for comments.

In fact now vcpkg uses WinHTTP in Windows, not curl. And it's a problem.
It would be more convenient to use curl in all the environments.

So I don't understand, why not to turn on WPAD in Windows?
In fact vcpkg works differently in Windows and Linux.

Really I have 2 patches for this exceptions.
One for Linux (with NO_PROXY environment variable), and one for Windows (with WPAD).

I'll publish PR with patches as soon as spossible.

**Comment by BillyONeal** [comment]
>In fact now vcpkg uses WinHTTP in Windows, not curl. And it's a problem.

If there's one header we need, for example, that falls back to curl.

>It would be more convenient to use curl in all the environments.

Hmmm... I argue the reverse. If we could use winhttp everywhere, that would eliminate a dependency. In particular, the smaller Windows Server builds for containers do not have a curl.exe.

>So I don't understand, why not to turn on WPAD in Windows?

If we could use the WinHTTP backend for everything on Windows, then we could turn WPAD on. The problem is that we often fall back to curl, and curl does not implement WPAD. It would be bad if half of vcpkg's operations used WPAD and the other half did not in a way over which users have no control.

That's why I said this above:

> > We could consider supporting WPAD, but this would require being able to use WinHTTP as the backend for all download operations.

If we never need the curl backend on Windows, then we can embrace this WinHTTP feature.

>Really I have 2 patches for this exceptions.
>One for Linux (with NO_PROXY environment variable), and one for Windows (with WPAD).

Can you explain more about how `NO_PROXY` is supposed to do anything here given that `NO_PROXY` (`WINHTTP_ACCESS_TYPE_NO_PROXY`) is exactly what we are already doing?

The intended behavior is that for the cases we choose WinHTTP, the behavior is indistinguishable from had we chosen curl. If you are seeing differences between WinHTTP and curl, that is a bug and we should fix that bug. But just flipping on WPAD is not a fix for that bug.

**Comment by AndreyAlifanov** [comment]
_Can you explain more about how NO_PROXY is supposed to do anything here given that NO_PROXY (WINHTTP_ACCESS_TYPE_NO_PROXY) is exactly what we are already doing?_

It does not work. At all. In our environment we've got errors with proxy without WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY.
These errors are in my initial post.

Ok, I see your point of view. I'll try to handle this problem with NO_PROXY environment variable. To ensure compatibility with curl.

**Comment by BillyONeal** [comment]
> By the way, WINHTTP_ACCESS_TYPE_NO_PROXY and NO_PROXY are absolutely different entities.
> NO_PROXY defines list of internal resources. Name resolution mechanism doesn't use proxy for them, but uses proxy for all other resources.
> WINHTTP_ACCESS_TYPE_NO_PROXY turns off proxy at all - for internal and external resources.

I see, that's the difference then. I'm guessing in your environment you have WPAD set up in a way that happens to configure things similarly to how you were setting `NO_PROXY`. We try to wire up HTTP(S)_PROXY to WinHTTP but apparently didn't do so for `NO_PROXY`.

It seems like the right fix is one of:

* Implement `NO_PROXY` handling for how we pass `HTTP(S)_PROXY` to WinHTTP, or
* Implement headers and similar handling so that we can always use WinHTTP on Windows which will let us turn on WPAD, or
* Add a switch to force use of curl

**Comment by BillyONeal** [comment]
Related: https://github.com/microsoft/vcpkg-tool/pull/1598

**Comment by demian51** [comment]
Hi everyone,

I ran into an issue on Windows where vcpkg takes my Linux-style `no_proxy` value (for example `localhost,127.0.0.1`) and passes it straight to WinHttp.

The problem is WinHTTP only understands semicolons, so any commas make the bypass list invalid and downloads fail.

If I change it to `localhost;127.0.0.1`, WinHTTP works fine but then curl and other tools break because they all expect commas. Since vcpkg also uses curl for some of its HTTP operations, those calls break too when given a semicolon-delimited list.

Is this verbatim passthrough to WinHTTP the intended behavior on Windows? Would it make sense for vcpkg to automatically convert commas to semicolons or split on both `,` and `;` and rejoin with `;` so that both WinHTTP and the cross-platform curl calls keep working?

**Comment by BillyONeal** [comment]
Part of the issue is that I don't use / understand these proxy settings myself; I was happy to merge code that seemed reasonable that folks who do use those said fixed things for them. The behavior hasn't been exactly 'designed' as a result.

A PR that changes , to ; to make WinHTTP may be reasonable?

**Comment by demian51** [comment]
Thanks for the clarification!

I'll open a PR that, on Windows, preprocesses the `NO_PROXY` value by replacing commas (`,`) with semicolons (`;`) in a local copy and passes that to `WinHttpSetOption`.

**Comment by demian51** [comment]
I did some more testing and it turns out that comma vs. semicolon alone wasn't the root cause in my environment. Both separators actually worked fine. CIDR entry (e.g. `10.0.0.0/8`) was the real culprit. I also discovered in the WinHTTP docs that the proxy bypass list must be delimited by semicolons or whitespace, not commas.

Given that official requirement, do we still want a PR that, on Windows, preprocesses commas into semicolons for convenience, or would it be better to simply update the documentation (and maybe add a note about CIDR entries) to reflect the delimiter and format rules?

Alternatively, would it make sense to extend the existing curl fallback logic (used when proxy auth or custom headers are required) to also kick in when the `NO_PROXY` value contains a CIDR expression?

**Comment by BillyONeal** [comment]
I'm a bit nervous to get into the business of trying to document how those env vars work because we are at most just a passthrough to the underlying HTTP implementation that actually implements the things. Trying to avoid being in the firing line for parsing etc. of web soup is part of why we shell out to curl in the first place.

So I would be a bit ... nervous to open the can of worms of trying to determine if it's CIDR.

**Comment by demian51** [comment]
I agree that parsing complex `NO_PROXY` values is out of scope.

What if we skip WinHTTP entirely when `NO_PROXY` is set and simply fall back to curl, instead of detecting CIDR notation?

**Comment by BillyONeal** [comment]
Related: https://github.com/microsoft/vcpkg-tool/pull/1660
